ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml

```diff
@@ -237,57 +237,68 @@ let dss_vcpus xc doms =
       in
       (* Runstate info is per-domain rather than per-vcpu *)
       let dss =
-        try
-          let ri = Xenctrl.domain_get_runstate_info xc domid in
-          ( Rrd.VM uuid
-          , Ds.ds_make ~name:"runstate_fullrun" ~units:"(fraction)"
-              ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time0 /. 1.0e9))
-              ~description:"Fraction of time that all VCPUs are running"
-              ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-          )
-          :: ( Rrd.VM uuid
-             , Ds.ds_make ~name:"runstate_full_contention" ~units:"(fraction)"
-                 ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time1 /. 1.0e9))
-                 ~description:
-                   "Fraction of time that all VCPUs are runnable (i.e., \
-                    waiting for CPU)"
-                 ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-             )
-          :: ( Rrd.VM uuid
-             , Ds.ds_make ~name:"runstate_concurrency_hazard"
-                 ~units:"(fraction)"
-                 ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time2 /. 1.0e9))
-                 ~description:
-                   "Fraction of time that some VCPUs are running and some are \
-                    runnable"
-                 ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-             )
-          :: ( Rrd.VM uuid
-             , Ds.ds_make ~name:"runstate_blocked" ~units:"(fraction)"
-                 ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time3 /. 1.0e9))
-                 ~description:
-                   "Fraction of time that all VCPUs are blocked or offline"
-                 ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-             )
-          :: ( Rrd.VM uuid
-             , Ds.ds_make ~name:"runstate_partial_run" ~units:"(fraction)"
-                 ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time4 /. 1.0e9))
-                 ~description:
-                   "Fraction of time that some VCPUs are running, and some are \
-                    blocked"
-                 ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-             )
-          :: ( Rrd.VM uuid
-             , Ds.ds_make ~name:"runstate_partial_contention"
-                 ~units:"(fraction)"
-                 ~value:(Rrd.VT_Float (Int64.to_float ri.Xenctrl.time5 /. 1.0e9))
-                 ~description:
-                   "Fraction of time that some VCPUs are runnable and some are \
-                    blocked"
-                 ~ty:Rrd.Derive ~default:false ~min:0.0 ()
-             )
-          :: dss
-        with _ -> dss
+        let dom_cpu_time =
+            Int64.(
+              to_float @@ logand dom.Xenctrl.cpu_time xen_flag_complement
```

Comment:
Why is this needed? This is collected per cpu in the precious lines

Reply:
I noticed the per vcpu retains this special process, I am not sure whether the per domain will encounter this issue or not so I kept it, if we can ensure per domain is ok then I will delete it.